Publish types for plugins package

noahtallen opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
I'm publishing types for more packages to avoid having to maintain stuff on DT.

## Why?
See #49647

## How?
Add the tsconfig and fix anything that comes up! Also check final build-types directory.

## Testing Instructions
CI

noahtallen committed:
Publish types for plugins package

noahtallen commented on packages/plugins/src/api/index.ts:
What do we think about changing some files to TS? Otherwise, we have to live with type issues in the file

tyxla replied:
I don't see a reason not to. Isn't there a reason to not want to include renaming to `ts` or `tsx` as part of the process to type a package?

dmsnell replied:
cc: @adamziel

it's been my experience that I think this move is hard _because_ it requires that our types check, which they usually don't. that means either removing a lot of JSDoc or making widespread changes, changes which can overwhelm and swamp a much simpler PR like this one.

personally I'd rather see us as a project eliminate DenitelyTyped types if we can, get mostly-useless-but-working types in Gutenberg, and then slowly refine the types here until we can change to TS without error.

this is all moot if there are no errors in this paackage

adamziel replied:
If you're able to make a switch to TS without creating a gazzillion new type errors, I'd say go for it. @dmsnell and I found it was quite difficult in other packages, but maybe this one would would give in easily?

As for the general strategy, I echo what Dennis said.

noahtallen replied:
> personally I'd rather see us as a project eliminate DenitelyTyped types if we can, get mostly-useless-but-working types in Gutenberg, and then slowly refine the types here until we can change to TS without error.

I agree.

This will be harder in other packages. We can turn off `checkJs` to silence issues with jsdoc types in JS files, which does make migrations easier. As an example, block-editor has 4888 errors if we just start checking everything. But none with `checkJs: false, skipLibCheck: true`. And this ultimately gives us full types for the store object, which is more than what the DT package provides, since DT store types are completely broken right now.

But like you said in another PR, we'd potentially be publishing incorrect types.

Though to be completely fair, they _are_ the types Gutenberg is using internally.

noahtallen replied:
At the same time, we have very extensive JSDoc types, so imo that plus inference gives us a lot more than mostly useless types.

So i guess it all comes down to whether it'll work to disable `checkJs`

noahtallen commented on packages/plugins/src/components/plugin-context/index.tsx:
```diff
@@ -15,13 +25,13 @@ export { Provider as PluginContextProvider };
+ *                          expected to return object of props to
+ *                          merge with the component's own props.
  *
- * @return {WPComponent} Enhanced component with injected context as props.
```
This `WPComponent` return type doesn't play nice -- way easier to let inference do its magic

noahtallen commented on packages/is-shallow-equal/src/index.js:
```diff
@@ -13,10 +13,10 @@ export { default as isShallowEqualArrays } from './arrays';
+ * otherwise. Also handles primitives and falsey values, just in case.
  *
- * @param {any[]|ComparableObject} a First object or array to compare.
- * @param {any[]|ComparableObject} b Second object or array to compare.
```
This function does already work with anything, and it's useful to allow it to receive a falsey value

tyxla replied:
If we're changing this, we likely should change it here as well:

https://github.com/WordPress/gutenberg/blob/a109fda769bfc3e33dc6dfd7c454afb9c8c60fb8/packages/is-shallow-equal/src/objects.js#L4-L5

noahtallen replied:
I slightly disagree -- the object comparison is only called after we know that the arguments are objects. I think that function does expect them to be objects. This top-level function decides which comparison function to use based on the input type -- if the input is anything other than an array or object, it'll fall back to strict equal, which is what allows primitives to work as input

tyxla commented on packages/is-shallow-equal/src/index.js:
```diff
@@ -13,10 +13,10 @@ export { default as isShallowEqualArrays } from './arrays';
 /**
  * Returns true if the two arrays or objects are shallow equal, or false
- * otherwise.
+ * otherwise. Also handles primitives and falsey values, just in case.
```
Falsy values are also primitives, aren't they?

```suggestion
 * otherwise. Also handles primitive values, just in case.
```

tyxla commented on packages/plugins/src/api/index.ts:
```diff
@@ -5,37 +5,47 @@
+	 * element (or function returning an element) if you choose to render your
+	 * own SVG.
+	 */
+	icon?: string | WPElement | Function;
```
I think this could be more precise, `string` and `Function` being too broad. 

We could use `IconKey` (AKA `DashiconIconKey`) instead of just a `string`. Or we could even use `IconType` from `@wordpress/components` for the entire type, and have `icon`'s type be `IconType | null`.

noahtallen replied:
I would love to do this, but this is a great example of another rabbit hole. `IconType` isn't exported from wordpress components. So just looking into how to solve this type issue and make it perfect took another 20 minutes. 

It's valuable, I don't want to detract from that, but we probably don't have time to work on this project if we try to make every type perfect along the way. Mostly because I'm trying to cover a lot of ground relatively quickly

noahtallen replied:
The side problem is that it _shouldn't_ be hard to export a type from components, but there aren't any other types exported from there, so I don't have an example to follow for what's acceptable

noahtallen replied:
And another problem is that this introduces a dependency on wordpress components that isn't strictly required except for development, but we don't have any way to encode that since we don't allow dev dependencies in packages

tyxla commented on packages/plugins/src/api/index.ts:
```diff
@@ -5,37 +5,47 @@
+	/**
+	 * A component containing the UI elements to be rendered.
+	 */
+	render: Function;
```
Should this be `React.ComponentType`? `Function` feels a bit broad.

tyxla commented on packages/plugins/src/api/index.ts:
```diff
@@ -5,37 +5,47 @@
- * @property {string}                    [scope] The optional scope to be used when rendering inside
- *                                               a plugin area. No scope by default.
  */
+export interface WPPlugin {
```
I wonder if we should try naming those as close as possible to their DT counterparts, to reduce the migration friction as much as possible.

tyxla commented on packages/plugins/src/api/index.ts:
```diff
@@ -5,37 +5,47 @@
- * @param {string}                 name     A string identifying the plugin.Must be
- *                                          unique across all registered plugins.
- * @param {Omit<WPPlugin, 'name'>} settings The settings for this plugin.
+ * @param name     A string identifying the plugin.Must be
```
```suggestion
 * @param name     A string identifying the plugin. Must be
```

tyxla commented on packages/plugins/src/api/index.ts:
```diff
@@ -126,7 +139,10 @@ export function registerPlugin( name, settings ) {
+	settings = applyFilters( 'plugins.registerPlugin', settings, name ) as Omit<
+		WPPlugin,
+		'name'
+	>;
```
`Omit< WPPlugin, 'name' >` gets used more than once here. Should we declare it inline as a separate type for brevity and DRY?

tyxla commented on packages/plugins/src/api/index.ts:
```diff
@@ -202,23 +218,23 @@ export function unregisterPlugin( name ) {
+ * @return The list of plugins without a scope or for a given scope.
  */
-export function getPlugins( scope ) {
+export function getPlugins( scope: string ) {
```
Previously we could omit the `scope` argument and that would be just fine. We even have unit tests that call it that way. The function will also work just fine if it's `undefined`. 

So, should we make that argument optional?

tyxla commented on packages/plugins/src/components/plugin-context/index.tsx:
```diff
@@ -4,7 +4,17 @@
+ */
+import type { WPPlugin } from '../../api';
+
+interface PluginContext {
+	name: null | WPPlugin[ 'name' ];
+	icon: null | WPPlugin[ 'icon' ];
+}
```
Should we just use an `Omit` type that omits the `render` prop?

Do those `null` defaults also mean that `null` should be a valid value for `name` and `icon`?

noahtallen replied:
Null is just the default value for the plugin context; it doesn't exactly mean that the plugin itself shouldn't have a name or icon.

I'm not super inclined to change existing types very much -- this is just encoding what plugin context already is

noahtallen replied:
I also don't think we should omit, because the context is hardcoded to just include icon and name and nothing else

tyxla commented on packages/plugins/src/components/plugin-context/index.tsx:
```diff
@@ -15,13 +25,13 @@ export { Provider as PluginContextProvider };
+ * @return Enhanced component with injected context as props.
  */
-export const withPluginContext = ( mapContextToProps ) =>
+export const withPluginContext = ( mapContextToProps: Function ) =>
```
Could we make that `Function` type more precise? As we can see down there on L41, it expects certain arguments (`context` and `props`).

tyxla commented on packages/plugins/tsconfig.json:
```diff
@@ -0,0 +1,17 @@
+		"rootDir": "src",
+		"declarationDir": "build-types",
+		"types": [ "gutenberg-env" ],
+		"checkJs": false
```
But we do have some JS files still (the error boundary) - do we really want to disable error reporting for JS files?

noahtallen replied:
Well, I agree it's definitely not ideal. But I think there are two parts of the project:

1. Let typescript infer everything it can from JSDoc, and publish those types for consumers.
2. Migrate to typescript (or at least turn on full type checking), solving every type issue along the way.

My ultimate goal is number 1. Number 2 would be amazing too, but it can _really_ severely impact the amount of work required. In something like data-controls, where there's only a single file, it's very easy.

But the real problem is that some of these packages are absolutely massive, and have thousands of type errors. block-editor is a better example where that would be impossible to do as part of this project.

I don't think we can increase the scope of this work to complete type safety in every one of these projects. My main goal is to get published types to replace DT asap, and I think that can be done without fixing every TS error in these existing projects. 

I also don't think it makes sense to block publishing build types on having completely correct internal types. The DT types don't enforce that, for example, and they are already broken for many packages. So publishing the build-types should result in a much better situation, even before every internal type issue is resolved.

noahtallen replied:
Plugins isn't the best example of all this, so I can take another look and see if it wouldn't bee too much work to turn this on.

But you raise a good point and I wanted to discuss it in the context of the broader project, especially since the remaining packages like editor and block-editor are significantly larger

noahtallen replied:
(rich-text as well for that matter)

tyxla reviewed (commented):
Great work @noahtallen 🙌 

I we could aim to get this closer to the existing DT types, at least when it comes to precision. I've left some suggestions, let me know what you think.

Also, let's not forget to add some CHANGELOG entries for a breaking change where applicable (in the plugins package most certainly).

noahtallen committed:
Update doc comment

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

noahtallen commented:
Another fantastic example of a rabbit hole: In order to fix all the type errors and start checking JS, we need to convert the `PluginArea` component to .tsx. Not too big a task, but after doing so, you can't commit, due to this error:

```
"Error: Could not find corresponding parameter token for documented parameter 'props.scope' in function 'PluginArea'.",
```

I did a fair amount of digging yesterday, and couldn't figure out why this _one_ thing causes an issue in this file, but not others. The solution is to keep the types in the docblock, instead of removing them. Definitely not ideal :/

Beyond that, the `npm run build:package-types` command will continue outputting errors in unrelated packages like components, and even dependencies like reakit and popper. Who knows why, but when disabling checkJS, it goes away :/

```shnode_modules/@popperjs/core/lib/types.d.ts(122,34): error TS2344: Type 'Options' does not satisfy the constraint 'Obj'.
node_modules/@popperjs/core/lib/types.d.ts(123,39): error TS2344: Type 'Options' does not satisfy the constraint 'Obj'.
node_modules/reakit/ts/Form/__utils/types.d.ts(14,5): error TS2411: Property '["0"]' of type 'keyof T | undefined' is not assignable to 'number' index type 'string | number'.
node_modules/reakit/ts/Form/__utils/types.d.ts(15,5): error TS2411: Property '["1"]' of type '(P extends { 0: infer K0; } ? K0 extends keyof T ? keyof T[K0] : never : never) | undefined' is not assignable to 'number' index type 'string | number'.
node_modules/reakit/ts/Form/__utils/types.d.ts(18,5): error TS2411: Property '["2"]' of type '(P extends { 0: infer K0; 1: infer K1; } ? K0 extends keyof T ? K1 extends keyof T[K0] ? keyof T[K0][K1] : never : never : never) | undefined' is not assignable to 'number' index type 'string | number'.
node_modules/reakit/ts/Form/__utils/types.d.ts(22,5): error TS2411: Property '["3"]' of type '(P extends { 0: infer K0; 1: infer K1; 2: infer K2; } ? K0 extends keyof T ? K1 extends keyof T[K0] ? K2 extends keyof T[K0][K1] ? keyof T[K0][K1][K2] : never : never : never : never) | undefined' is not assignable to 'number' index type 'string | number'.
node_modules/reakit/ts/Form/__utils/types.d.ts(27,5): error TS2411: Property '["4"]' of type '(P extends { 0: infer K0; 1: infer K1; 2: infer K2; 3: infer K3; } ? K0 extends keyof T ? K1 extends keyof T[K0] ? K2 extends keyof T[K0][K1] ? K3 extends keyof T[K0][K1][K2] ? keyof T[K0][K1][K2][K3] : never : never : never : never : never) | undefined' is not assignable to 'number' index type 'string | number'.
node_modules/reakit/ts/Form/__utils/types.d.ts(33,5): error TS2411: Property '["5"]' of type '(P extends { 0: infer K0; 1: infer K1; 2: infer K2; 3: infer K3; 4: infer K4; } ? K0 extends keyof T ? K1 extends keyof T[K0] ? K2 extends keyof T[K0][K1] ? K3 extends keyof T[K0][K1][K2] ? K4 extends keyof T[K0][K1][K2][K3] ? keyof T[K0][K1][K2][K3][K4] : never : never : never : never : never : never) | undefined' is not assignable to 'number' index type 'string | number'.
packages/components/build-types/view/component.d.ts(24,28): error TS2314: Generic type 'WordPressComponent' requires 3 type argument(s).
husky - pre-commit hook exited with code 1 (error)
```

noahtallen committed:
Convert more things to TS and clarify types

noahtallen committed:
Update plugins changelog

noahtallen commented:
I also don't think the components change justifies a changelog entry here :p

noahtallen commented:
I wasn't sure if those errors would just occur locally, but it does look like they spill over here. Would appreciate help figuring them out since they seem completely unrelated :/

dmsnell commented on packages/is-shallow-equal/src/index.js:
```diff
@@ -13,10 +13,10 @@ export { default as isShallowEqualArrays } from './arrays';
- * @param {any[]|ComparableObject} a First object or array to compare.
- * @param {any[]|ComparableObject} b Second object or array to compare.
+ * @param {any} a First object or array to compare.
+ * @param {any} b Second object or array to compare.
```
not to make this more confusing, but isn't `unknown` possibly more useful here than `any`? whereas `any` satisfies all type constraints, the `unknown` would only widen as we make programmatic checks to assert its type.

noahtallen replied:
Can do!

tyxla commented:
> Another fantastic example of a rabbit hole: In order to fix all the type errors and start checking JS, we need to convert the `PluginArea` component to .tsx. Not too big a task, but after doing so, you can't commit, due to this error:
> 
> ```
> "Error: Could not find corresponding parameter token for documented parameter 'props.scope' in function 'PluginArea'.",
> ```
> 
> I did a fair amount of digging yesterday, and couldn't figure out why this _one_ thing causes an issue in this file, but not others. The solution is to keep the types in the docblock, instead of removing them. Definitely not ideal :/
> 
> Beyond that, the `npm run build:package-types` command will continue outputting errors in unrelated packages like components, and even dependencies like reakit and popper. Who knows why, but when disabling checkJS, it goes away :/
> 
> ```
> node_modules/@popperjs/core/lib/types.d.ts(123,39): error TS2344: Type 'Options' does not satisfy the constraint 'Obj'.
> node_modules/reakit/ts/Form/__utils/types.d.ts(14,5): error TS2411: Property '["0"]' of type 'keyof T | undefined' is not assignable to 'number' index type 'string | number'.
> node_modules/reakit/ts/Form/__utils/types.d.ts(15,5): error TS2411: Property '["1"]' of type '(P extends { 0: infer K0; } ? K0 extends keyof T ? keyof T[K0] : never : never) | undefined' is not assignable to 'number' index type 'string | number'.
> node_modules/reakit/ts/Form/__utils/types.d.ts(18,5): error TS2411: Property '["2"]' of type '(P extends { 0: infer K0; 1: infer K1; } ? K0 extends keyof T ? K1 extends keyof T[K0] ? keyof T[K0][K1] : never : never : never) | undefined' is not assignable to 'number' index type 'string | number'.
> node_modules/reakit/ts/Form/__utils/types.d.ts(22,5): error TS2411: Property '["3"]' of type '(P extends { 0: infer K0; 1: infer K1; 2: infer K2; } ? K0 extends keyof T ? K1 extends keyof T[K0] ? K2 extends keyof T[K0][K1] ? keyof T[K0][K1][K2] : never : never : never : never) | undefined' is not assignable to 'number' index type 'string | number'.
> node_modules/reakit/ts/Form/__utils/types.d.ts(27,5): error TS2411: Property '["4"]' of type '(P extends { 0: infer K0; 1: infer K1; 2: infer K2; 3: infer K3; } ? K0 extends keyof T ? K1 extends keyof T[K0] ? K2 extends keyof T[K0][K1] ? K3 extends keyof T[K0][K1][K2] ? keyof T[K0][K1][K2][K3] : never : never : never : never : never) | undefined' is not assignable to 'number' index type 'string | number'.
> node_modules/reakit/ts/Form/__utils/types.d.ts(33,5): error TS2411: Property '["5"]' of type '(P extends { 0: infer K0; 1: infer K1; 2: infer K2; 3: infer K3; 4: infer K4; } ? K0 extends keyof T ? K1 extends keyof T[K0] ? K2 extends keyof T[K0][K1] ? K3 extends keyof T[K0][K1][K2] ? K4 extends keyof T[K0][K1][K2][K3] ? keyof T[K0][K1][K2][K3][K4] : never : never : never : never : never : never) | undefined' is not assignable to 'number' index type 'string | number'.
> packages/components/build-types/view/component.d.ts(24,28): error TS2314: Generic type 'WordPressComponent' requires 3 type argument(s).
> husky - pre-commit hook exited with code 1 (error)
> ```

> I wasn't sure if those errors would just occur locally, but it does look like they spill over here. Would appreciate help figuring them out since they seem completely unrelated :/

I haven't had a chance to take a look at these errors myself, but this might indicate that there are existing TS errors in the components package that for some reason are manifesting only when we start explicitly exporting types. cc @ciampo @mirka and @chad1008 in case they have any idea what might be going on.

mirka reviewed (commented):
> Beyond that, the `npm run build:package-types` command will continue outputting errors in unrelated packages like components, and even dependencies like reakit and popper. Who knows why, but when disabling checkJS, it goes away :/

I looked into this, and it's due to an [upstream issue](https://github.com/WordPress/gutenberg/pull/34373#issuecomment-908242950) in Reakit. It's already been [fixed](https://github.com/ariakit/ariakit/commit/710c2e8dce370a4e5c8401f417ca4ba64341f549) in reakit@1.3.9, so I'll see if I can update our dependency version asap (currently 1.3.8). The quick fix is to add `"skipLibCheck": true` in the `plugins/tsconfig.json`.

noahtallen committed:
Resolve review comments

noahtallen commented:
Alright, updated the PR:

- Added a components changelog entry
- Added skipLibCheck to avoid the error 
- Changed `isShallowEqual` argument types to unknown
- Started typechecking JS

adamziel commented on packages/plugins/src/components/plugin-area/index.tsx:
```diff
@@ -16,15 +16,27 @@ import isShallowEqual from '@wordpress/is-shallow-equal';
- * @param {Function|undefined} props.onError
+ * @param                                                    props
+ * @param {string}                                           props.scope
+ * @param {( name: Plugin[ 'name' ], error: Error ) => void} props.onError
```
Oh, this syntax works – nice

adamziel replied:
Hold on, this is a `tsx` file now – shouldn't the inline type `PluginAreaProps` listed below with the arguments suffice?

tyxla replied:
My thoughts exactly. We should be able to remove much of this existing JSDoc.

noahtallen replied:
@adamziel @tyxla You're 100% right, and I already did try to do just that (three times now, just to be sure)... but unfortunately the documentation linter is very unhappy with this file, and I couldn't get it to work. So I needed to copy the types here :( 

I spent enough time on it that I don't think it's worth trying to fix, but happy for one of you to take a look :) 

```sh
Error: Command failed with exit code 1: "/Users/noahallen/source/gutenberg/node_modules/.bin/docgen" packages/plugins/src/index.js --output packages/plugins/README.md --to-token --use-token "Autogenerated API docs" --ignore "/unstable|experimental/i"

Error: Could not find corresponding parameter token for documented parameter 'props.scope' in function 'PluginArea'.
  at makeError (/Users/noahallen/source/gutenberg/node_modules/execa/lib/error.js:59:11)
  at handlePromise (/Users/noahallen/source/gutenberg/node_modules/execa/index.js:114:26)
  at processTicksAndRejections (internal/process/task_queues.js:95:5)
  at async Transform.<anonymous> (/Users/noahallen/source/gutenberg/bin/api-docs/update-api-docs.js:219:5) {
shortMessage: 'Command failed with exit code 1: "/Users/noahallen/source/gutenberg/node_modules/.bin/docgen" packages/plugins/src/index.js --output packages/plugins/README.md --to-token --use-token "Autogenerated API docs" --ignore "/unstable|experimental/i"',
command: '"/Users/noahallen/source/gutenberg/node_modules/.bin/docgen" packages/plugins/src/index.js --output packages/plugins/README.md --to-token --use-token "Autogenerated API docs" --ignore "/unstable|experimental/i"',
exitCode: 1,
```

noahtallen replied:
see #49812

adamziel replied:
Great catch

dmsnell commented on packages/plugins/README.md:
```diff
@@ -147,12 +147,12 @@ registerPlugin( 'plugin-name', {
 _Returns_
 
--   `WPPlugin`: The final plugin settings object.
+-   The final plugin settings object.
```
we've lost this return type here, I'm guessing because we lost the return type annotation in the function declaration. in any case, it seems like the existing JSDoc type was wrong since it's `WPPlugin | null` according to what I see.

dmsnell commented on packages/plugins/src/api/index.ts:
```diff
@@ -105,9 +118,9 @@ const plugins = {};
+ * @return The final plugin settings object.
  */
-export function registerPlugin( name, settings ) {
+export function registerPlugin( name: string, settings: PluginSettings ) {
```
as with the other functions here, I think it's likely a good idea that _if_ we remove a JSDoc return type annotation we ought to add it to the function signature. `export function registerPlugin( … ): WPPlugin | null { … }`

1. this can help expose type errors more easily that inference might pass through through widening.
2. we have lost the annotation in our generated docs (though I think those are honestly of little value anyway)

sometimes the return type annotation is significant for TS to figure out un-annotated types. I recently ran into this in Elm, then coincidentally saw an article discussing the phenomena in Rust as a leading reason why they require the full signatures. the problem for me appeared in the form of type errors appearing non-locally which are confusing and hard to track down.

noahtallen replied:
Great point!

tyxla commented on packages/components/src/index.ts:
```diff
@@ -90,6 +90,7 @@ export { default as GuidePage } from './guide/page';
 export { Heading as __experimentalHeading } from './heading';
 export { HStack as __experimentalHStack } from './h-stack';
 export { default as Icon } from './icon';
+export type { IconType } from './icon';
```
@ciampo @mirka are we planning to have any special or dedicated way to export types from the `@wordpress/components` package, or can we just export them as needed like this? 

Should we export all types at some point?

mirka replied:
> can we just export them as needed like this?

Yes, at the moment I think it'd be better to export like this when needed 🤔 We have tons of type aliases and I don't think we want to commit to them (or their names) being stable. We can revisit later if it's a problem.

tyxla commented on packages/components/CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@
 
 -   `DropZone`: Smooth animation ([#49517](https://github.com/WordPress/gutenberg/pull/49517)).
 -   `Navigator`: Add `skipFocus` property in `NavigateOptions`. ([#49350](https://github.com/WordPress/gutenberg/pull/49350)).
+-   Export `IconType` type for external use. ([#49649](https://github.com/WordPress/gutenberg/pull/49649))
```
Should this be in `New Feature` instead of `Enhancements`, since it's exposing a new type? Or should it be `Internal` since it's about types?

mirka replied:
I think enhancement is appropriate here. (Also in changelogs I personally advocate for semver-like groupings, in which "new feature" and "enhancement" are the same thing and do not need to be differentiated.)

tyxla replied:
Sounds good 👍

tyxla commented on packages/plugins/README.md:
```diff
@@ -180,19 +180,19 @@ _Parameters_
 _Returns_
 
--   `WPPlugin | undefined`: The previous plugin settings object, if it has been successfully unregistered; otherwise `undefined`.
+-   The previous plugin settings object, if it has been successfully unregistered; otherwise `undefined`.
```
This is similar to what @dmsnell said above about losing the type.

tyxla commented on packages/plugins/README.md:
```diff
@@ -180,19 +180,19 @@ _Parameters_
 
 _Returns_
 
--   `WPComponent`: Enhanced component with injected context as props.
```
Same as above with regards to losing the type.

noahtallen replied:
This one is harder to fix, because none of these types (`WPElement`, `WPComponent`, `JSX.Element`, `React.ComponentType`, `React.Element`) are compatible with what typescript infers from `createHigherOrderComponent`. So it's a lot easier to just let inference do its magic here. (I'm also not convinced that having the "WPComponent" text in the docs really provides enough meaningful value to try to fix this issue :/)

noahtallen replied:
I take that back; apparently, we can just put WPComponent in JSdoc, and it's happy enough 😅

noahtallen committed:
Restore return types

noahtallen committed:
Update package-lock file

noahtallen committed:
Inline argument type and update docs

adamziel commented on packages/plugins/src/components/plugin-area/index.tsx:
```diff
@@ -16,15 +16,22 @@ import isShallowEqual from '@wordpress/is-shallow-equal';
- * @param {string|undefined}   props.scope
- * @param {Function|undefined} props.onError
+ * @param  props
+ * @param  props.scope
```
Do you have to list all the properties now that they are statically typed? AFAIR other places just say `@param props`

noahtallen replied:
My IDE automatically inserts them on save, so I think yes?

adamziel reviewed (approved):
I left one more nit but otherwise this is awesome!

tyxla commented on packages/components/CHANGELOG.md:
```diff
@@ -4,10 +4,13 @@
 -   `TreeGrid`: Modify keyboard navigation code to use a data-expanded attribute if aria-expanded is to be controlled outside of the TreeGrid component ([#48461](https://github.com/WordPress/gutenberg/pull/48461)).
 -   `Modal`: Equalize internal spacing ([#49890](https://github.com/WordPress/gutenberg/pull/49890)).
 -   `Modal`: Increased border radius ([#49870](https://github.com/WordPress/gutenberg/pull/49870)).
--   `SlotFill`: Added util for creating private SlotFills and supporting Symbol keys ([#49819](https://github.com/WordPress/gutenberg/pull/49819)).
```
This looks like it might have been caused by a bad rebase. We should likely revert it.

tyxla commented on packages/components/CHANGELOG.md:
```diff
@@ -4,10 +4,13 @@
 
+-   Export `IconType` type for external use. ([#49649](https://github.com/WordPress/gutenberg/pull/49649))
+
+### Enhancements
```
We'll need to move this accordingly since the Enhancements section already exists.

tyxla reviewed (approved):
Looking great from my perspective 🚀 

Left a couple of last-minute CHANGELOG suggestions, good to go after they're addressed.

noahtallen committed:
Fix changelog

ciampo commented:
Do y'all think that this PR (and the related ones publishing types) deserve a dev note?

tyxla commented:
> Do y'all think that this PR (and the related ones publishing types) deserve a dev note?

I think that is a great idea. @noahtallen is this something you're planning to work on? I'm happy to collaborate there.

noahtallen commented:
## 6.3 Dev Note:

Typescript developers appreciate having great in-editor type-hints and compiler warnings. We would also like to reduce our reliance on DefinitelyTyped, as it's much harder to maintain type information separately from the source code. As a result, we are publishing more types for various `@wordpress` packages by allowing Typescript to emit types by parsing our JSDoc comments.

With some small tweaks to assist its inference capabilities, Typescript can generate extensive type definitions even for store objects without migrating much code to Typescript. So far, we have published types for the following packages using this process:

- rich-text (https://github.com/WordPress/gutenberg/pull/49651)
- data-controls (https://github.com/WordPress/gutenberg/pull/49647)
- notices (https://github.com/WordPress/gutenberg/pull/49650)
- plugins (https://github.com/WordPress/gutenberg/pull/49649)

With work in progress on some larger packages such as `blocks`, `editor`, and `block-editor`.
